### PR TITLE
SLCORE-813 Keep track of the user telemetry in the VS context

### PR DESCRIPTION
**Pull request by kirill-knize-sonarsource**

### CODE REVIEW

File: API_CHANGES.md [hunk 1]
```diff
@@ -80,6 +80,9 @@
 * Add a method to the `org.sonarsource.sonarlint.core.rpc.protocol.backend.analysis.AnalysisRpcService` to let the client notify the backend with user defined analysis properties
   * `org.sonarsource.sonarlint.core.rpc.protocol.backend.analysis.AnalysisRpcService#didSetUserAnalysisProperties` to set user defined properties
 * For analysis, both user-defined and inferred properties will be merged. If the same property is inferred by the client and provided by the user - the inferred value will be used for analysis.
+* Add a method to `org.sonarsource.sonarlint.core.rpc.protocol.SonarLintRpcClient` to allow the backend to request client-defined file
```
Comment (vojtech-suchy-sonarsource) [inline API_CHANGES.md:83]: Forgotten changes?

File: rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/backend/initialize/InitializeParams.java [hunk 2]
```diff
@@ -88,6 +89,7 @@ public InitializeParams(
    * @param sonarlintUserHome           Path to SonarLint user home directory. If null, will default to ~/.sonarlint
    * @param standaloneRuleConfigByKey   Local rule configuration for standalone analysis. This configuration will override defaults rule activation and parameters.
    */
+  @Deprecated(since = "10.4")
```
Comment (vojtech-suchy-sonarsource) [inline rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/backend/initialize/InitializeParams.java:92]: Should we create a ticket to go through the deprecated methods during hardening? We now have the third very large constructor in this class.

**Review by vojtech-suchy-sonarsource — APPROVED** [review]
Overall looks good.  See minor comments.

Now we have two mechanisms for migrating telemetry 
- one for migrating the old telemetry by parsing file
-  one for which lets clients to provide data through InitParams.

Just noting that I would prefer extending the first approach as now we will be forced to have the new TelemetryMigrationDto which is practically  used only by VS.  But regarding the current capacity and the upcoming release schedule I guess we should go with that.

File: API_CHANGES.md (hunk 1 above)
Comment (kirill-knize-sonarsource) [reply API_CHANGES.md]: Yes, well spotted. Removed. Also, documentation for the new init param was added.

File: rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/backend/initialize/InitializeParams.java (hunk 2 above)
Comment (kirill-knize-sonarsource) [reply rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/backend/initialize/InitializeParams.java:92]: Good point. Created: https://sonarsource.atlassian.net/browse/SLCORE-914

**Pull request merged**